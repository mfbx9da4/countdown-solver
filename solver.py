import random


def add(a, b): return a+b


def sub(a, b): return a-b


def mul(a, b): return a*b


def div(a, b): return a/b if a % b == 0 else 0/0


operations = [(add, '+'),
              (sub, '-'),
              (mul, '*'),
              (div, '/')]


def OneFromTheTop():
    return [25, 50, 75][random.randint(0, 2)]


def OneOfTheOthers():
    return random.randint(1, 10)


def Evaluate(stack):
    try:
        total = 0
        lastOper = add
        for item in stack:
            if type(item) is int:
                total = lastOper(total, item)
            else:
                lastOper = item[0]

        return total
    except:
        return 0


def ReprStack(stack):
    reps = [str(item) if type(item) is int else item[1] for item in stack]
    return ' '.join(reps)


def Solve(target, numbers):

    def Recurse(stack, nums):
        for n in range(len(nums)):
            stack.append(nums[n])

            remaining = nums[:n] + nums[n+1:]

            if Evaluate(stack) == target:
                print(ReprStack(stack))

            if len(remaining) > 0:
                for op in operations:
                    stack.append(op)
                    stack = Recurse(stack, remaining)
                    stack = stack[:-1]

            stack = stack[:-1]

        return stack

    Recurse([], numbers)


# target = random.randint(100, 1000)

# numbers = [OneFromTheTop()] + [OneOfTheOthers() for i in range(5)]

target, numbers = [356, [75, 6, 8, 5, 1, 10]]
# target, numbers = [4, [2, 2]]

print("Target: {0} using {1}".format(target, numbers))

Solve(target, numbers)
