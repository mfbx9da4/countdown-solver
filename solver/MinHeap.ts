import TinyQueue from 'tinyqueue'

export class MinHeap<T extends { priority: number }> {
  private heap = new TinyQueue([], (a, b) => a.priority - b.priority)
  push(node: T): void {
    // TODO: implement
    this.heap.push(node)
  }
  pop(): T {
    // TODO: implement
    return this.heap.pop()
  }

  get length(): number {
    return this.heap.length
  }
}
