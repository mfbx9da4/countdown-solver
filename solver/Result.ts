type ResultPathNode = {
  value: string
  partialOutput: number
}

export interface Result {
  output: number
  distance: number
  resultId: string
  path: ResultPathNode[]
}
