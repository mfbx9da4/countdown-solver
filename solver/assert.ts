class AssertionError extends Error {
  public extra: any
  constructor(message, extra) {
    super(message)
    this.name = 'AssertionError'
    this.message = message
    this.extra = extra
  }
}

const defaultOpts = { log: true }

function assert(
  predicate: any,
  message: string,
  extra: any = {},
  opts: Partial<typeof defaultOpts> = defaultOpts
): asserts predicate {
  if (!predicate) {
    const error = new AssertionError(message, extra)
    if (opts.log) {
      // eslint-disable-next-line no-console
      console.log('AssertionError:', message, extra)
      console.error(error)
    }
    throw error
  }
}

export default assert
