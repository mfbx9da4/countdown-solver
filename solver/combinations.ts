export interface Combination<T> {
  combination: T[]
  remaining: T[]
}

function format<T>(indexes: number[], arr: T[]) {
  const combination: T[] = []
  const remaining: T[] = []
  const indexesSet = new Set(indexes)
  for (let i = 0; i < arr.length; i++) {
    const item = arr[i]
    if (indexesSet.has(i)) combination.push(item)
    else remaining.push(item)
  }

  return { combination, remaining }
}

function rec<T>(arr: T[], choices: number, lo = 0): number[][] {
  let ret: number[][] = []
  if (choices === 0) return ret
  if (choices === 1) {
    for (let i = lo; i < arr.length; i++) ret.push([i])
    return ret
  }
  for (let i = lo; i < arr.length; i++) {
    for (const subResult of rec(arr, choices - 1, i + 1)) {
      ret.push([i, ...subResult])
    }
  }
  return ret
}

export function combinations<T>(arr: T[], choices: number): Combination<T>[] {
  return rec(arr, choices).map((x) => format(x, arr))
}

export interface Partition<T> {
  left: T[]
  right: T[]
}

export function partitions<T>(arr: T[]): Partition<T>[] {
  if (arr.length <= 1) return []
  const ret: Partition<T>[] = []
  for (let i = 1; i < arr.length - 1; i++) {
    for (const combination of combinations(arr, i)) {
      ret.push({ left: combination.combination, right: combination.remaining })
      ret.push({ left: combination.remaining, right: combination.combination })
    }
  }
  return ret
}
