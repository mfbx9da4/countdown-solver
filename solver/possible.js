const fac = (n) => !n || n * fac(n - 1)

const possible = (n) => {
  let s = 0
  for (let i = 1; i < n + 1; i++) {
    s += fac(i) * 4 ** (i - 1)
  }
  return s
}

console.log('possible(6)', possible(6))
