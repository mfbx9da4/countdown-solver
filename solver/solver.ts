const ops = ['*', '/', '+', '-'] as const
type Op = typeof ops[number]
const open = '('
const close = ')'
type Paren = typeof open | typeof close
type Expression = Array<Op | Paren | number>

interface Node {
  expression: Expression
  needsOp: boolean
  remaining: number[]
  /** distance to the target */
  priority: number
}

interface Result {
  expression: Expression
  output: number
  distance: number
}

class MinHeap<T extends { priority: number }> {
  private heap: T[] = []
  push(node: T) {
    // TODO: implement
    this.heap.push(node)
  }
  pop() {
    // TODO: implement
    return this.heap.pop()
  }

  get length() {
    return this.heap.length
  }
}

// function closeTrailingParens(node: Node) {
//   let balance = node.balance
//   const expression = [...node.expression]
//   while (balance > 0) {
//     expression.push(close)
//     balance--
//   }
//   return expression
// }

function parenthesize(expression: Expression): Expression[] {
  if (expression.length === 3) return [[open, ...expression, close]]
  if (expression.length < 3) return [expression]
  const ret: Expression[] = []
  ret.push(expression)
  for (let size = expression.length - 2; size > 2; size -= 2) {
    for (let i = 0; i <= expression.length - size; i += 2) {
      for (const before of parenthesize(expression.slice(0, i))) {
        for (const middle of parenthesize(expression.slice(i, i + size))) {
          for (const end of parenthesize(expression.slice(i + size))) {
            ret.push([...before, ...middle, ...end])
          }
        }
      }
    }
  }
  return ret
}

// const arg = [25, '+', 25, '+', 75]
// for (const p of parenthesize(arg)) {
//   console.log('r', p.join(''))
// }

function evaluate(expression: Expression): number {
  try {
    return eval(expression.join(''))
  } catch (e) {
    console.log('failed to evaluate', expression.join('')) // eslint-disable-line
    return NaN
  }
}

/**
 * goal is to explore every permutation of the inputs.
 * We consider these unique permutations:
 *   3 - (4 + 5)
 *   3 - (5 + 4)
 *   3 - 4 - 5
 * or even these are considered unique
 *   3 + 3
 *   3 + 3
 * Even though mathematically they are the same expression rearranged.
 * We go through every permutation of numbers and every permutation of parens
 * the root is `(0)+(`
 * balance is `(` - `)`
 *
 *
 * much simpler approach is to do as I originally thought and two steps
 * one which build number op, number op
 * second which does paren permutations
 *
 *
 *
 * at every even step we can choose from one of the ops
 * at every odd step we can choose from:
 *  either: one of the remaining numbers
 *  or: an opening paren if the balance is < inputs.length + 1
 *  or: a closing paren if the balance is > 0 and prev not open paren
 * after every even step we can emit the expression
 * in evaluate we strip out empty () and add trailing close parens
 * and eval
 * we add the children to the min heap with a priority
 * priority will be difference to the target
 * a priority 0 is better
 * all evaluated expressions are added to a separate min heap
 * Enhancements:
 *  * add leading minus (-4 - 5) + 3
 *  * discard zero sums
 *  * handle non integer division
 *  * handle zero division
 *  * Remove unnecessary brackets
 *  * The order of the same number shouldn't matter
 *  * Under addition or multiplication order should not matter
 *  * Smart parens - not needed if associative
 */
export function* solve(
  target: number,
  inputs: number[]
): Generator<Result, Result[], Result> {
  if (!inputs.length) throw new Error('no inputs')
  const root: Node = {
    expression: [],
    needsOp: false,
    priority: target,
    remaining: inputs,
  }
  const results: Result[] = []
  const heap = new MinHeap<Node>()
  heap.push(root)
  while (heap.length) {
    const node = heap.pop()

    // console.log('node.state', node)

    if (node.needsOp) {
      let distance = Infinity
      for (const expression of parenthesize(node.expression)) {
        const output = evaluate(expression)
        distance = Math.abs(target - output)
        if (isNaN(output)) continue
        const result: Result = { expression, output, distance }
        results.push(result)
        yield result
      }

      if (node.remaining.length !== 0) {
        // add an op
        for (const op of ops) {
          const child: Node = {
            ...node,
            expression: [...node.expression, op],
            priority: distance,
            needsOp: false,
          }
          heap.push(child)
        }
      }
    } else {
      for (let i = 0; i < node.remaining.length; i++) {
        // pick a number from remaining
        const num = node.remaining[i]
        const remaining = [
          ...node.remaining.slice(0, i),
          ...node.remaining.slice(i + 1),
        ]
        const child: Node = {
          ...node,
          remaining,
          expression: [...node.expression, num],
          needsOp: true,
        }
        heap.push(child)
      }
    }
  }
  return results
}
