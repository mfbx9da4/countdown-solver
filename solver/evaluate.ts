import { Expression, evaluatePair, Op } from './solver'

export function evaluate(expression: Expression): number {
  try {
    let ret = 0
    let lastOp = '+'
    for (let i = 0; i < expression.length; i += 2) {
      const num = expression[i]
      ret = evaluatePair(ret, num as number, lastOp as Op)
      lastOp = expression[i + 1] as Op
    }
    return ret
  } catch (e) {
    console.log('failed to evaluate', expression.join('')) // eslint-disable-line
    return NaN
  }
}
