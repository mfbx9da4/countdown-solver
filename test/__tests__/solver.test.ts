import { solve as solveIter } from '../../solver/solver'
import assert from 'assert'

const solve = (...args: Parameters<typeof solveIter>) => [...solveIter(...args)]

describe('solver', () => {
  it('base case', () => {
    const results = solve(1, [1])
    assert.strictEqual(results.length, 1, 'too many')
    const [result] = results
    assert.strictEqual(result.distance, 0, 'bad')
    expect(results).toMatchSnapshot()
  })

  it('2 permutations', () => {
    const results = solve(4, [1, 2])
    assert.strictEqual(results.length, 8, 'too many')
    const result = results.find((x) => x.distance === 1)
    assert.ok(result.distance, 'did not find closest')
    expect(results).toMatchSnapshot()
  })

  it('3 permutations', () => {
    const results = solve(5, [4, 2, 3])
    // console.log('results', results)
    assert.strictEqual(results.length, 8, 'too many')
    const result = results.find((x) => x.distance === 1)
    assert.ok(result.distance, 'did not find closest')
    expect(results).toMatchSnapshot()
  })
})
