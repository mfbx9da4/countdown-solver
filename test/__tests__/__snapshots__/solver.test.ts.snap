// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`solver 2 permutations 1`] = `
Array [
  Object {
    "distance": 3,
    "expression": Array [
      1,
    ],
    "output": 1,
  },
  Object {
    "distance": 2,
    "expression": Array [
      1,
      "*",
      2,
    ],
    "output": 2,
  },
  Object {
    "distance": 1,
    "expression": Array [
      1,
      "+",
      2,
    ],
    "output": 3,
  },
  Object {
    "distance": 2,
    "expression": Array [
      2,
    ],
    "output": 2,
  },
  Object {
    "distance": 2,
    "expression": Array [
      2,
      "*",
      1,
    ],
    "output": 2,
  },
  Object {
    "distance": 3,
    "expression": Array [
      2,
      "-",
      1,
    ],
    "output": 1,
  },
  Object {
    "distance": 1,
    "expression": Array [
      2,
      "+",
      1,
    ],
    "output": 3,
  },
  Object {
    "distance": 2,
    "expression": Array [
      2,
      "/",
      1,
    ],
    "output": 2,
  },
]
`;

exports[`solver all permutations 1`] = `
Array [
  Object {
    "distance": 3,
    "expression": Array [
      1,
    ],
    "output": 1,
  },
  Object {
    "distance": 2,
    "expression": Array [
      1,
      "*",
      2,
    ],
    "output": 2,
  },
  Object {
    "distance": 1,
    "expression": Array [
      1,
      "+",
      2,
    ],
    "output": 3,
  },
  Object {
    "distance": 2,
    "expression": Array [
      2,
    ],
    "output": 2,
  },
  Object {
    "distance": 2,
    "expression": Array [
      2,
      "*",
      1,
    ],
    "output": 2,
  },
  Object {
    "distance": 3,
    "expression": Array [
      2,
      "-",
      1,
    ],
    "output": 1,
  },
  Object {
    "distance": 1,
    "expression": Array [
      2,
      "+",
      1,
    ],
    "output": 3,
  },
  Object {
    "distance": 2,
    "expression": Array [
      2,
      "/",
      1,
    ],
    "output": 2,
  },
]
`;

exports[`solver base case 1`] = `
Array [
  Object {
    "distance": 0,
    "expression": Array [
      1,
    ],
    "output": 1,
  },
]
`;
